### PR TITLE
Feature/path like support for file trait

### PR DESCRIPTION
**Pull request by midhun-pm**
Working towards #642

### CODE REVIEW

File: traits/trait_types.py [hunk 1]
```diff
@@ -1430,6 +1431,9 @@ def validate(self, object, name, value):
 
             Note: The 'fast validator' version performs this check in C.
         """
+        if isinstance(value, PathLike):
+            value = fspath(value)
```
Comment (corranwebster) [inline traits/trait_types.py]: If you give `os.fspath` a string it will return it unmodified: https://docs.python.org/3/library/os.html#os.fspath

So it is better to  just call it in a `try: ... except TypeError: ...`, I think.  if the exception happens then it doesn't validate, otherwise you have a string which is going to hold a filesystem path.

**Comment by mdickinson** [comment]
@midhun-pm As a general rule, please could you avoid using force push for PRs that have already been reviewed? It makes it harder to keep track of the history of the comments and reviews.

Don't worry about the eventual history in Traits master: we use squash-and-merge, so everything gets squashed down to a single commit anyway.

File: traits/trait_types.py [hunk 2]
```diff
@@ -1430,6 +1431,21 @@ def validate(self, object, name, value):
 
             Note: The 'fast validator' version performs this check in C.
         """
+        try:
+            # Note: Python 3.5 does not implement __fspath__
```
Comment (mdickinson) [inline traits/trait_types.py]: I'd be happy to implement this functionality *just* for Python 3.6 and later, and for Python 3.5 keep the current `File = Str` behaviour. I suspect we have very few users who care about Python 3.5.

File: traits/trait_types.py [hunk 3]
```diff
@@ -1430,6 +1431,21 @@ def validate(self, object, name, value):
...
+            from os import fspath
```
Comment (mdickinson) [inline traits/trait_types.py]: I think it would be better to move the `try` / `except` for the `ImportError` to the top of the module, and define a "_HAVE_OS_FSPATH" constant (a better name almost certainly exists). Doing a failed import on every validation is unnecessarily expensive.

File: traits/trait_types.py [hunk 4]
```diff
@@ -1387,7 +1388,7 @@ class BaseFile(BaseStr):
     """
 
     #: A description of the type of value this trait accepts:
-    info_text = "a file name"
+    info_text = "a file name or PathLike object"
```
Comment (mdickinson) [inline traits/trait_types.py]: Suggest "instance of os.PathLike" instead of "PathLike object", since `PathLike` doesn't get instantiated directly. "object implementing the os.PathLike interface" would work, too.

File: traits/tests/test_file.py [hunk 5]
```diff
@@ -9,7 +9,10 @@
 # Thanks for using Enthought open source!
 
 import os
+import sys
+
 import unittest
+from pathlib import Path
```
Comment (mdickinson) [inline traits/tests/test_file.py]: Let's follow the usual Enthought rules for import order here: group all std. lib. imports together (so lose the blank line between `sys` and `unittest`, and keep the modules in alphabetical order.

File: traits/tests/test_file.py [hunk 6]
```diff
@@ -36,6 +43,15 @@ def assign_invalid():
...
         self.assertRaises(TraitError, assign_invalid)
 
+    @unittest.skipIf(sys.version_info < (3, 6), "PathLike File trait test")
+    def test_invalid_pathlike_file(self):
+        example_model = ExampleModel(file_name=__file__)
+
+        def assign_invalid():
+            example_model.file_name = Path("not_valid_path!#!#!#")
+
+        self.assertRaises(TraitError, assign_invalid)
```
Comment (mdickinson) [inline traits/tests/test_file.py]: I'd suggest using the context manager form of `self.assertRaises` here. That way you don't need a separate function.
```
with self.assertRaises(TraitError):
    example_model.file_name = ...
```

File: traits/tests/test_file.py [hunk 7]
```diff
@@ -44,6 +60,15 @@ def assign_invalid():
...
         self.assertRaises(TraitError, assign_invalid)
 
+    @unittest.skipIf(sys.version_info < (3, 6), "PathLike File trait test")
+    def test_pathlike_directory(self):
+        example_model = ExampleModel(file_name=__file__)
+
+        def assign_invalid():
+            example_model.file_name = Path(os.path.dirname(__file__))
+
+        self.assertRaises(TraitError, assign_invalid)
```
Comment (mdickinson) [inline traits/tests/test_file.py]: Same comment as above: this would be clearer with a `with`.

File: traits/tests/test_file.py [hunk 8]
```diff
@@ -52,6 +77,7 @@ def assign_invalid():
 
         self.assertRaises(TraitError, assign_invalid)
 
+    @unittest.skip("Unimplemented fast validator for File trait")
```
Comment (mdickinson) [inline traits/tests/test_file.py]: I'd suggest keeping this test, not skipping it. It should still pass.

File: traits/trait_types.py [hunk 9]
```diff
@@ -20,10 +20,15 @@
...
 import sys
+try:
+    from os import fspath
+except ImportError:
+    fspath = None
 from os.path import isfile, isdir
 from types import FunctionType, MethodType, ModuleType
 import uuid
 
+
```
Comment (mdickinson) [inline traits/trait_types.py]: Unrelated spacing change: please revert. (Our usual conventions are for only a single line between import blocks.)

File: traits/trait_types.py [hunk 10]
```diff
@@ -1467,14 +1482,14 @@ def __init__(
...
         """
-        if not exists:
-            # Define the C-level fast validator to use:
-            self.fast_validate = (ValidateTrait.coerce, str)
 
         super(File, self).__init__(
             value, filter, auto_set, entries, exists, **metadata
         )
 
+    def validate(self, object, name, value):
```
Comment (mdickinson) [inline traits/trait_types.py]: Can this method simply be removed? It looks to me as though it should be fine to simply allow `File` to inherit the `validate` implementation of `BaseFile`.

**Comment by midhun-pm** [comment]
Thanks for the review! Made those changes.
